api/sixtyfps-node/native/js_model.rs

```diff
@@ -64,22 +64,27 @@ impl Model for JsModel {
         r.get()
     }
 
-    fn row_data(&self, row: usize) -> Self::Data {
-        let r = Cell::new(sixtyfps_interpreter::Value::default());
-        crate::run_with_global_context(&|cx, persistent_context| {
-            let row = JsNumber::new(cx, row as f64);
-            let obj = self.get_object(cx, persistent_context).unwrap();
-            let _ = obj
-                .get(cx, "rowData")
-                .ok()
-                .and_then(|func| func.downcast::<JsFunction>().ok())
-                .and_then(|func| func.call(cx, obj, std::iter::once(row)).ok())
-                .and_then(|res| {
-                    crate::to_eval_value(res, self.data_type.clone(), cx, persistent_context).ok()
-                })
-                .map(|res| r.set(res));
-        });
-        r.into_inner()
+    fn row_data(&self, row: usize) -> Option<Self::Data> {
+        if row >= self.row_count() {
+            None
+        } else {
+            let r = Cell::new(sixtyfps_interpreter::Value::default());
+            crate::run_with_global_context(&|cx, persistent_context| {
+                let row = JsNumber::new(cx, row as f64);
+                let obj = self.get_object(cx, persistent_context).unwrap();
+                let _ = obj
+                    .get(cx, "rowData")
+                    .ok()
+                    .and_then(|func| func.downcast::<JsFunction>().ok())
+                    .and_then(|func| func.call(cx, obj, std::iter::once(row)).ok())
+                    .and_then(|res| {
+                        crate::to_eval_value(res, self.data_type.clone(), cx, persistent_context)
+                            .ok()
+                    })
+                    .map(|res| r.set(res));
+            });
+            Some(r.into_inner())
```

Comment:
Similar to the C++ suggestion below, for JavaScript we should probably change the `Model` interface in `index.ts` to use `rowData(row: number): T | undefined;` and allow the JS model to do the check.

Reply:
I can look into this, but the tests @ogoffart suggested work without this, so I'll keep that for an extra patch. This one is too big already.

Reply:
my tests are written in Javascript, not TypeScript

Yeah, we don't have typescript tests :-s

Reply:
> I can look into this, but the tests @ogoffart suggested work without this, so I'll keep that for an extra patch. This one is too big already.

Sounds good.

The `write_to_model.60` test in the `js` section calls for example `assert.equal(another_model.rowData(2).score, 439.);`.

If we changed `rowData` also in JS to return `undefined` for an out-of-bounds row, then this could be extended here.